dev/ci/go-test-backcompat.sh

```diff
@@ -0,0 +1,125 @@
+      num_changed=$((num_changed + 1))
+    fi
+  done < <(find "${1}" -name '*_test.go' -type f)
+
+  if [ ${num_changed} -eq 0 ]; then
+    echo "Unknown test in ${1}: ${2}"
+  fi
+}
+
+# `disable_test_path ${path} ${prefix}`
+function disable_test_path() {
+  echo "Disabling test '${2}*' in ${1}"
+
+  if [ -d "${1}" ]; then
+    disable_test_dir "${1}" "${2}"
+  elif [ -f "${1}" ]; then
+    disable_test_file "${1}" "${2}"
+  fi
+}
+
+current_head=$(git rev-parse HEAD)
+latest_minor_release_tag="v${MINIMUM_UPGRADEABLE_VERSION}"
+latest_minor_release_commit=$(git merge-base "${current_head}" "${latest_minor_release_tag}")
+flakefile="./dev/ci/backcompat-flake/${latest_minor_release_tag}.json"
+
+# Early exit
+if git diff --quiet "${latest_minor_release_commit}".."${current_head}" migrations; then
+  echo "--- No schema changes"
+  echo "No schema changes since last minor release"
+  exit 0
+fi
+
+echo "--- Running backwards compatibility tests"
+echo "current_head                = ${current_head}"
+echo "latest_minor_release_tag    = ${latest_minor_release_tag}"
+echo "latest_minor_release_commit = ${latest_minor_release_commit}"
+echo ""
+echo "Running Go tests to test database schema backwards compatibility:"
+echo "- database schemas are defined at HEAD / ${current_head}, and"
+echo "- unit tests are defined at the last minor release ${latest_minor_release_tag} / ${latest_minor_release_commit}."
+echo ""
+echo "New failures of these tests indicate that new changes to the"
+echo "database schema do not allow for continued proper operation of"
+echo "Sourcegraph instances deployed at the previous release."
+echo ""
+
+PROTECTED_FILES=(
+  ./migrations
+  ./dev/ci/backcompat-flake/
+  ./dev/ci/go-test.sh
+)
+
+# Check out the previous code then immediately restore whatever
+# the current version of the protected files are.
+git checkout "${latest_minor_release_commit}"
+git checkout "${current_head}" -- "${PROTECTED_FILES[@]}"
+
+if [ -f "${flakefile}" ]; then
+  echo ""
+  echo "Disabling tests listed in flakefile ${flakefile}"
+
+  for pair in $(jq -r '.[] | "\(.path):\(.prefix)"' <"${flakefile}"); do
+    IFS=' ' read -ra parts <<<"${pair/:/ }"
+    disable_test_path "${parts[0]}" "${parts[1]}"
+  done
+fi
+
+# Re-run asdf to ensure we have the correct set of utilities to
+# run the currently checked out version of the Go unit tests.
+./.buildkite/hooks/pre-command
+
+if ! ./dev/ci/go-test.sh "$@"; then
+  echo ""
+  echo "!!! This commit contains database schema definitions that caused an"
+  echo "unexpected failure of one or more unit tests at tagged commit ${latest_minor_release_tag}."
+  echo ""
+  echo "If this backwards incompatibility is intentional or of the test is flaky,"
+  echo "an exception for this text can be added to the following flakefile:"
+  echo ""
+  echo "'${flakefile}'"
+  echo ""
+  echo "Strongly consider rewriting these changes to make a backwards compatible"
+  echo "change instead."
+  exit 1
```

Comment:
```suggestion
  echo "change instead."
  echo ""
  echo "If you opt to make this a backwards incompatible change, ... TODO"
  exit 1
```

Maybe changelog item or something - I'm not sure how you might want this handled (or is it not important to call out?)

Reply:
I'll be writing docs to link here pre-merge.

Reply:
https://github.com/sourcegraph/sourcegraph/pull/29389